CleverTapSDK/Encryption/CTCryptMigrator.m

```diff
@@ -0,0 +1,307 @@
+        return NO;
+    }
+}
+
+#pragma mark - In-app Migration
+
+- (BOOL)migrateInAppData {
+    return ([self migrateInAppsWithKeySuffix:CLTAP_PREFS_INAPP_KEY] &&
+            [self migrateInAppsWithKeySuffix:CLTAP_PREFS_INAPP_KEY_CS] &&
+            [self migrateInAppsWithKeySuffix:CLTAP_PREFS_INAPP_KEY_SS]);
+}
+
+- (BOOL)migrateInAppsWithKeySuffix:(NSString *)keySuffix {
+    if (!keySuffix || keySuffix.length == 0) {
+        CleverTapLogInfo(self.config.logLevel, @"Error: Key suffix is nil or empty.");
+        return NO;
+    }
+    
+    NSString *key = [self storageKeyWithSuffix:keySuffix];
+    if (!key) {
+        CleverTapLogInfo(self.config.logLevel, @"Error: Failed to generate storage key.");
+        return NO;
+    }
+    
+    id value = [CTPreferences getObjectForKey:key];
+    
+    if (!value) {
+        CleverTapLogInfo(self.config.logLevel, @"Warning: No value found for key: %@", key);
+        return YES; // No value to migrate, but not a failure
+    }
+
+    NSString *encryptedString = nil;
+    
+    if ([value isKindOfClass:[NSString class]]) {
+        encryptedString = (NSString *)value;
+    } else if ([value isKindOfClass:[NSArray class]]) {
+        NSArray *arrayValue = (NSArray *)value;
+        if (arrayValue.count > 0 && [arrayValue.firstObject isKindOfClass:[NSString class]]) {
+            encryptedString = arrayValue.firstObject;
+        } else {
+            CleverTapLogInfo(self.config.logLevel, @"Error: Array value is empty or contains non-string elements.");
+            return NO;
+        }
+    } else {
+        CleverTapLogInfo(self.config.logLevel, @"Error: Unsupported value type: %@", [value class]);
+        return NO;
+    }
+
+    if (!encryptedString) {
+        CleverTapLogInfo(self.config.logLevel, @"Error: Encrypted string is nil.");
+        return NO;
+    }
+
+    if (![_cryptManager isTextAESGCMEncrypted:encryptedString]) {
+        NSArray *arr = [_cryptManager decryptObject:encryptedString encryptionAlgorithm:AES];
+        if (!arr) {
+            CleverTapLogInfo(self.config.logLevel, @"Error: Decryption failed for string: %@", encryptedString);
+            return NO;
+        }
+        
+        NSString *migratedEncryptedString = [_cryptManager encryptObject:arr];
+        if (!migratedEncryptedString) {
+            CleverTapLogInfo(self.config.logLevel, @"Error: Encryption failed after decryption.");
+            return NO;
+        }
+        NSString *newStorageKey = [self storageKeyWithSuffix:keySuffix];
+        [CTPreferences putString:migratedEncryptedString forKey:newStorageKey];
+        
+        CleverTapLogInfo(self.config.logLevel, @"GUID migration completed successfully for key: %@", keySuffix);
+    } else {
+        CleverTapLogInfo(self.config.logLevel, @"Value is already AES-GCM encrypted, no migration needed for key: %@", keySuffix);
+    }
+
+    return YES;
+}
+
+- (NSString *)storageKeyWithSuffix:(NSString *)suffix {
```

Comment:
Duplicate code from the in-app store, check if this can be shared.
It is also not clear what this storage key is for in the current context.

Reply:
we already have multiple variations of `storageKeyWithSuffix` in different classes based on different inputs. Its difficult to make a generic one since the inputs and their counts are different everytime.

Reply:
`storageKeyWithSuffix` in a corresponding class refers to the storage key this class uses to store data. However, in this case, the `storageKeyWithSuffix refers` to the in-app store storage keys. This is why I have requested it to be renamed in this case.
I see that it has been renamed.